### PR TITLE
ZRC-2: Fungible Token

### PR DESCRIPTION
**Pull request by GreyEcologist**

### CODE REVIEW

File: zrcs/zrc-2.md [hunk 1]
```diff
@@ -0,0 +1,372 @@
+| ZRC | Title                            | Status | Type     | Author                             | Created (yyyy-mm-dd) | Updated (yyyy-mm-dd) |
+| --- | -------------------------------- | ------ | -------- | ---------------------------------- | -------------------- | -------------------- |
+| 1   | Standard for Fungible Tokens | Draft  | Ready | Gareth Mensah <gareth@zilliqa.com> | 2019-11-18           | 2019-11-18           |
+
+## I. What are Fungible Tokens?
+
+The Fungible Token is an open standard for creating currencies. Fungibility is the property of a good or a commodity whose individual units are essentially interchangeable, and each of its parts is indistinguishable from another part.
```
Comment (vaivaswatha) [inline zrcs/zrc-2.md]: I'm not sure if using "good" is good (not good good :wink: ). https://english.stackexchange.com/questions/421860/is-the-word-good-as-the-singular-of-goods-correct.

Maybe rephrase as "Fungibility is the property of goods or commodities whose individual units are ..."

File: zrcs/zrc-2.md [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+
+## II. Abstract
+
+ZRC-2 defines a minimum interface a smart contract must implement to allow fungible tokens to be managed, tracked, owned, and traded.
```
Comment (vaivaswatha) [inline zrcs/zrc-2.md]: defines a minimum interface **that** a smart contract ...

File: zrcs/zrc-2.md [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+3. the transitions that will allow the changing of values of the mutable variables;
+4. the events to be emitted by them.
+
+### A. Roles
+
+| Name              | Description                                                                                                                                                                                                                      |     |
+| ----------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |--|
+| `contractOwner`   | The owner of the contract initialized by the creator of the contract.                                                                                                                                                            |
+| `tokenOwner`      | A user (identified by an address) that owns a token tied to a tokenId.                                                                                                                                                           |
```
Comment (vaivaswatha) [inline zrcs/zrc-2.md]: Fungible tokens doesn't have the concept of `tokenId` right? Looks like this has come in why copying over from non-fungible tokens.

File: zrcs/zrc-2.md [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+| @param | `amount`  | `Uint128` | Number of tokens to be burned.           |
+
+|           | Name          | Description                | Event Parameters                                                                                                                                                                                                                                               |
+| --------- | ------------- | -------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| eventName | `BurnSuccess` | Burning is successful.     | `from`: `ByStr20`, and `amount`: `Uint128`.                                                                                                                      |
+| eventName | `Error`       | Burning is not successful. | - emit `CodeNotAuthorised` if the transition is called by a user who is not the contract owner.<br>**NOTE:** Only the `tokenOwner` is allowed to call this transition. |
```
Comment (vaivaswatha) [inline zrcs/zrc-2.md]: There is a bit of confusion here. The first sentence says that only the contract owner can call, but the second one says that the `tokenOwner` should call.

File: zrcs/zrc-2.md [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+```ocaml
+(* @dev: Burn existing tokens. Only tokenOwner or approved operator can burn a token *)
+(* @param from:                             Address holding the tokens to be burned. *)
+(* @param amount:                           Number of tokens to be destroyed.        *)
+transition Burn(from: ByStr20, amount: Uint128)
+```
+
+|        | Name      | Type      | Description                              |
+| ------ | --------- | --------- | ---------------------------------------- |
+| @param | `from`    | `ByStr20` | Address holding the tokens to be burned. |
```
Comment (vaivaswatha) [inline zrcs/zrc-2.md]: maybe `from` can be renamed to be `burnee` or `burn_account`.

File: zrcs/zrc-2.md [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+```ocaml
+(* @dev: Mint new tokens. Only contractOwner can mint.                 *)
+(* @param to:     Address of the recipient whose balance is increased. *)
+(* @param amount: Number of tokens to be burned.                       *)
+transition Mint(to: ByStr20, amount: Uint128)
+```
+
+|        | Name      | Type      | Description                                          |
+| ------ | --------- | --------- | ---------------------------------------------------- |
+| @param | `to`      | `ByStr20` | Address of the recipient whose balance is increased. |
```
Comment (vaivaswatha) [inline zrcs/zrc-2.md]: `to` -> `recipient` ?

File: zrcs/zrc-2.md (hunk 4 above)
Comment (GreyEcologist) [reply zrcs/zrc-2.md]: Yes, it should say token owner.

File: zrcs/zrc-2.md (hunk 5 above)
Comment (GreyEcologist) [reply zrcs/zrc-2.md]: burn_account it is!

File: zrcs/zrc-2.md [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
+transition DefaultOperators()
+```
+
+|           | Name                       | Description                             | Event Parameters                                                                                                                                                                                                               |
+| --------- | -------------------------- | --------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
+| eventName | `DefaultOperatorsSuccess` | Listing default operators is successful.     | TBA. |
+| eventName | `Error`                    | Listing default operators is not successful. | TBA.                                                                                                      |
+
+
+#### 11. Transfer
```
Comment (vaivaswatha) [inline zrcs/zrc-2.md]: I'm wondering if this should be called `Send` and the `Send` above should be called `Transfer`.

Comment (GreyEcologist) [reply zrcs/zrc-2.md]: I don't disagree that it would make more sense, but then there's also the `TransferFrom` transition. 

The potentially confusing part of using the ERC-777 standard is that it introduce operators but also supports the ERC-20 nomenclature. I merely followed the naming of their methods. I think for the sake of staying aligned with the ERC-777 naming convention, at least for now, we should stick the existing naming?

I could create a divide in the ZRC, indicating which part is the operator flow vs the erc20 flow?

File: zrcs/zrc-2.md [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+(* There can only be one approved spender per token at a given time         *)
+(* param tokenHolder:  Address of a token holder.                           *)
+(* param spender:      Address to be set as a spender.                      *)
+(* param amount:       Number of tokens to be approved for a given spender. *)
+transition Approve(tokenHolder: ByStr20, spender: ByStr20, amount: Uint128)
+```
+
+|        | Name      | Type      | Description                                          |
+| ------ | --------- | --------- | ---------------------------------------------------- |
+| @param | `amount`  | `ByStr20` | Address of a token owner.                            |
```
Comment (vaivaswatha) [inline zrcs/zrc-2.md]: `amount` here must be replaced with `tokenOwner`

Comment (GreyEcologist) [reply zrcs/zrc-2.md]: Got it. I've replaced `tokenHolder` with `tokenOwner` throughout.

File: zrcs/zrc-2.md [hunk 9]
```diff
@@ -0,0 +1,412 @@
...
+| eventName | `OperatorBurnSuccess` | Burning is successful.     | `operator`: `ByStr20`, `tokenOwner`: `ByStr20`, and `amount`: `Uint128`.                                                                                               |
+| eventName | `Error`          | Burning is not successful. | - emit `CodeNotAuthorised` if the transition is called by an operator who is not authorized. |
+
+
+#### 7. Mint
+
+```ocaml
+(* @dev: Mint new tokens. Only contractOwner can mint.                        *)
+(* @param recipient:     Address of the recipient whose balance is increased. *)
+(* @param amount:        Number of tokens to be burned.                       *)
```
Comment (vaivaswatha) [inline zrcs/zrc-2.md]:
```suggestion
(* @param amount:        Number of tokens to be minted.                       *)
```

File: zrcs/zrc-2.md [hunk 10]
```diff
@@ -0,0 +1,412 @@
...
+| eventName | `Error`       | Minting is not successful. | - emit `CodeTokenExists` if the token already exists.<br>- emit `CodeNotAuthorised` if the transition is called by a user who is not the contract owner.<br>**NOTE:** Only the `contractOwner` is allowed to call this transition. |
+
+
+#### 8. OperatorMint
+
+```ocaml
+(* @dev: Mint new tokens. Only approved operator can mint tokens.         *)
+(* @param operator:   Address must be an operator of tokenOwner.          *)
+(* @param recipient: Address of the recipient whose balance is increased. *)
+(* @param amount:    Number of tokens to be burned.                       *)
```
Comment (vaivaswatha) [inline zrcs/zrc-2.md]:
```suggestion
(* @param amount:    Number of tokens to be minted.                       *)
```

File: zrcs/zrc-2.md [hunk 11]
```diff
@@ -0,0 +1,412 @@
...
+| --------- | ------------- | -------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| eventName | `MintSuccess` | Minting is successful.     | `recipient`: `ByStr20`, and `amount`: `Uint128`.                             |
+| eventName | `Error`       | Minting is not successful. | - emit `CodeTokenExists` if the token already exists.<br>- emit `CodeNotAuthorised` if the transition is called by a user who is not the contract owner.<br>**NOTE:** Only the `contractOwner` is allowed to call this transition. |
+
+
+#### 8. OperatorMint
+
+```ocaml
+(* @dev: Mint new tokens. Only approved operator can mint tokens.         *)
+(* @param operator:   Address must be an operator of tokenOwner.          *)
```
Comment (vaivaswatha) [inline zrcs/zrc-2.md]: Similar to `OperatorSend`, the `operator` need not be a parameter here. It is just the `_sender`.

File: zrcs/zrc-2.md [hunk 12]
```diff
@@ -0,0 +1,412 @@
...
+(* @param recipient: Address of the recipient whose balance is increased. *)
+(* @param amount:    Number of tokens to be burned.                       *)
+transition OperatorMint(operator: ByStr20, recipient: ByStr20, amount: Uint128)
+```
+
+|        | Name         | Type      | Description                                         |
+| ------ | ------------ | --------- | --------------------------------------------------- |
+| @param | `operator`   | `ByStr20` | Address of a default operator.                      |
```
Comment (vaivaswatha) [inline zrcs/zrc-2.md]: `operator` shouldn't be a parameter

File: zrcs/zrc-2.md [hunk 13]
```diff
@@ -0,0 +1,412 @@
...
+| ------ | ---------- | --------- | ------------------------------ |
+| @param | `operator` | `ByStr20` | Address to be set as operator. Cannot be calling address. |
+
+|           | Name                       | Description                             | Event Parameters                                                                                                                                                                                                               |
+| --------- | -------------------------- | --------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
+| eventName | `AuthorizeOperatorSuccess` | Authorizing is successful.     | `operator`: `ByStr20`. |
+| eventName | `Error`                    | Authorizing is not successful. | - emit `CodeNotAuthorised` if the transition is called by a user who is not the token holder.                                                                                                      |
```
Comment (vaivaswatha) [inline zrcs/zrc-2.md]: I'm not sure I get the `Error` part. Anybody should be able to authorize anybody else as an operator right? Why should the error be "CodeNotAuthorized" ? There is no "token" to be considered here for someone to be "token holder".

File: zrcs/zrc-2.md [hunk 14]
```diff
@@ -0,0 +1,412 @@
...
+| ------ | ---------- | --------- | -------------------------------- |
+| @param | `operator` | `ByStr20` | Address to be unset as operator. |
+
+|           | Name                       | Description                             | Event Parameters                                                                                                                                                                                                               |
+| --------- | -------------------------- | --------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
+| eventName | `RevokeOperatorSuccess` | Revoking is successful.     | `operator`: `ByStr20`. |
+| eventName | `Error`                    | Revoking is not successful. | - emit `CodeNotAuthorised` if the transition is called by a user who is not the token holder.                                                                                                      |
```
Comment (vaivaswatha) [inline zrcs/zrc-2.md]: Same comment as for `AuthorizeOperator`.

File: zrcs/zrc-2.md [hunk 15]
```diff
@@ -0,0 +1,412 @@
...
+| ------ | ------------- | --------- | ------------------------------------ |
+| @param | `operator`    | `ByStr20` | Address of a potential operator.     |
+| @param | `tokenOwner`  | `ByStr20` | Address of a token ownwer.           |
+
+|           | Name                       | Description                             | Event Parameters                                                                                                                                                                                                               |
+| --------- | -------------------------- | --------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
+| eventName | `IsOperatorForSuccess` | Checking operator is successful.     | `operator`: `ByStr20`, and `tokenOwner`: `ByStr20`. |
+| eventName | `Error`                    | Checking operator is not successful. | TBA.                                                                                                      |
```
Comment (vaivaswatha) [inline zrcs/zrc-2.md]: I don't think there should be any error here at all. It is just a `True` or `False` result (event).

File: zrcs/zrc-2.md [hunk 16]
```diff
@@ -0,0 +1,412 @@
...
+```
+
+|        | Name          | Type      | Description                          |
+| ------ | ------------- | --------- | ------------------------------------ |
+| @param | `operator`    | `ByStr20` | Address of a potential operator.     |
+| @param | `tokenOwner`  | `ByStr20` | Address of a token ownwer.           |
+
+|           | Name                       | Description                             | Event Parameters                                                                                                                                                                                                               |
+| --------- | -------------------------- | --------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
+| eventName | `IsOperatorForSuccess` | Checking operator is successful.     | `operator`: `ByStr20`, and `tokenOwner`: `ByStr20`. |
```
Comment (vaivaswatha) [inline zrcs/zrc-2.md]: I think rather than the "checking operator" being successful or not, the event must be whether the `operator` is an operator or not, i.e., just a `True` or `False` event.

File: zrcs/zrc-2.md [hunk 17]
```diff
@@ -0,0 +1,412 @@
...
+
+|           | Name                  | Description                 | Event Parameters                                                                                                                                                                                                                                                                        |
+| --------- | --------------------- | --------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| eventName | `TransferSuccess` | Transfering is successful.     | `to`: `ByStr20`, and `amount`: `Uint128`.                                                                            |
+| eventName | `Error`               | Transfering is not successful. | - emit `CodeNotAuthorised` if the transition is called by a user that is not authorised.<br>**NOTE:** Only `tokenOwner` address can invoke this transition. |
```
Comment (vaivaswatha) [inline zrcs/zrc-2.md]: Again, since the `tokenOwner` is implicitly the caller, the error should more appropriately be "out of balance". The sender is _always_ authorized, he just may or may not have sufficient balance.

File: zrcs/zrc-2.md [hunk 18]
```diff
@@ -0,0 +1,412 @@
...
+(* @dev: Move a given amount of tokens from one address another using the allowance mechanism. *)
+(* param from:    Address of the sender whose balance is deccreased.                           *)
+(* param to:      Address of the recipient whose balance is increased.                         *)
+(* param amount:  Number of tokens to be transferred.                                          *)
+transition TansferFrom(from: ByStr20, to: ByStr20, amount: Uint128)
+```
+
+|        | Name      | Type      | Description                                          |
+| ------ | --------- | --------- | ---------------------------------------------------- |
+| @param | `from`    | `ByStr20` | Address of the sender whose balance is deccreased.   |
```
Comment (vaivaswatha) [inline zrcs/zrc-2.md]:
```suggestion
| @param | `from`    | `ByStr20` | Address of the sender whose balance is decreased.   |
```

File: zrcs/zrc-2.md [hunk 19]
```diff
@@ -0,0 +1,412 @@
...
+| --------- | --------------------- | --------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| eventName | `TransferSuccess` | Transfering is successful.     | `to`: `ByStr20`, and `amount`: `Uint128`.                                                                            |
+| eventName | `Error`               | Transfering is not successful. | - emit `CodeNotAuthorised` if the transition is called by a user that is not authorised.<br>**NOTE:** Only `tokenOwner` address can invoke this transition. |
+
+#### 14. TansferFrom
+
+```ocaml
+(* @dev: Move a given amount of tokens from one address another using the allowance mechanism. *)
```
Comment (vaivaswatha) [inline zrcs/zrc-2.md]:
```suggestion
(* @dev: Move a given amount of tokens from one address to another using the allowance mechanism. The caller must be an `approvedSpender`. *)
```

File: zrcs/zrc-2.md [hunk 20]
```diff
@@ -0,0 +1,412 @@
...
+
+|           | Name             | Description                 | Event Parameters                                                                                                                                                                                                                                              |
+| --------- | ---------------- | --------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| eventName | `TansferFromSuccess` | Approval is successful.     | `from`: `ByStr20`, `to`: `ByStr20`, and `amount`: `Uint128`.                                                                                               |
+| eventName | `Error`          | Approval is not successful. | - emit `CodeNotAuthorised` if the transition is called by a user who is not authorized to approve. <br>**NOTE:** Only either the `tokenOwner` or approved `operator`(s) are allowed to call this transition. |
```
Comment (vaivaswatha) [inline zrcs/zrc-2.md]:
```suggestion
| eventName | `Error`          | Approval is not successful. | - emit `CodeNotAuthorised` if the transition is called by a user who is not authorized to approve. <br>**NOTE:** Only either the `tokenOwner`, `approvedSpender` or authorized `operator`s are allowed to call this transition. |
```

File: zrcs/zrc-2.md [hunk 21]
```diff
@@ -0,0 +1,412 @@
...
+| @param | `amount`     | `Uint128` | Number of tokens to be approved for a given spender. |
+
+|           | Name             | Description                 | Event Parameters                                                                                                                                                                                                                                              |
+| --------- | ---------------- | --------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| eventName | `ApproveSuccess` | Approving is successful.     | `spender`: `ByStr20`, and `amount`: `Uint128`.                                                                                               |
+| eventName | `Error`          | Approving is not successful. | - emit `CodeNotAuthorised` if the transition is called by a user who is not authorized to approve. <br>**NOTE:** Only the `tokenOwner` or approved is allowed to call this transition. |
```
Comment (vaivaswatha) [inline zrcs/zrc-2.md]:
```suggestion
| eventName | `Error`          | Approving is not successful. | - emit `CodeNotAuthorised` if the transition is called by a user who is not authorized to approve. <br>**NOTE:** Only the `tokenOwner`  is allowed to call this transition. |
```

File: zrcs/zrc-2.md (hunk 13 above)
Comment (GreyEcologist) [reply zrcs/zrc-2.md]: @vaivaswatha The sender cannot make himself/herself the operator.

File: zrcs/zrc-2.md (hunk 14 above)
Comment (GreyEcologist) [reply zrcs/zrc-2.md]: @vaivaswatha if the operator is not found then it cannot be revoked.

File: zrcs/zrc-2.md (hunk 15 above)
Comment (GreyEcologist) [reply zrcs/zrc-2.md]: fixed.

File: zrcs/zrc-2.md (hunk 16 above)
Comment (GreyEcologist) [reply zrcs/zrc-2.md]: fixed
